src/SignalR/clients/ts/signalr/tests/AutomaticReconnect.test.ts

```diff
@@ -0,0 +1,449 @@
+
+            TestWebSocket.webSocket.onopen(new TestEvent());
+
+            await startPromise;
+
+            connection.onclose = (e) => {
+                closePromise.resolve();
+            };
+
+            connection.onreconnecting = (e) => {
+                onreconnectingCallCount++;
+                reconnectingPromise.resolve();
+            };
+
+            connection.onreconnected = (connectionId) => {
+                onreconnectedCalled = true;
+            };
+
+            TestWebSocket.webSocketSet = new PromiseSource();
+
+            TestWebSocket.webSocket.onclose(new TestCloseEvent());
+
+            await reconnectingPromise;
+
+            for (const _ of reconnectDelays) {
+                await TestWebSocket.webSocketSet;
+                await TestWebSocket.webSocket.closeSet;
+
+                TestWebSocket.webSocketSet = new PromiseSource();
+
+                TestWebSocket.webSocket.onerror(new TestEvent());
+            }
+
+            await closePromise;
+
+            expect(onreconnectingCallCount).toBe(1);
+            expect(onreconnectedCalled).toBe(false);
+
+            expect(negotiateCount).toBe(5);
+        },
+        "Connection disconnected with error 'Error: WebSocket closed with status code: 0 ().'.",
+        "Failed to start the transport 'WebSockets': null",
+        "Failed to start the transport 'ServerSentEvents': Error: 'ServerSentEvents' is disabled by the client.",
+        "Failed to start the transport 'LongPolling': Error: 'LongPolling' is disabled by the client.");
+    });
+
+    it("does not reset after hub handshake failure", async () => {
+        const reconnectDelays = [0, 0];
+        let negotiateCount = 0;
+
+        await VerifyLogger.run(async (logger) => {
+            const options = {
+                ...commonOptions,
+                httpClient: new TestHttpClient()
+                    .on("POST", () => {
+                        negotiateCount++;
+                        return defaultNegotiateResponse;
+                    }),
+                logger,
+                reconnectPolicy: new DefaultReconnectPolicy(reconnectDelays),
+            };
+
+            const reconnectingPromise = new PromiseSource();
+            const closePromise = new PromiseSource();
+            let onreconnectedCalled = false;
+            let onreconnectingCallCount = 0;
+
+            const connection = new HttpConnection("http://tempuri.org", options);
+            const hubConnection = HubConnection.create(connection, logger, new JsonHubProtocol());
+
+            TestWebSocket.webSocketSet = new PromiseSource();
+
+            const startPromise = hubConnection.start();
+
+            await TestWebSocket.webSocketSet;
+            await TestWebSocket.webSocket.closeSet;
+
+            TestWebSocket.webSocket.readyState = TestWebSocket.OPEN;
+            TestWebSocket.webSocket.onopen(new TestEvent());
+
+            await new Promise((resolve) => setTimeout(resolve, 100));
```

Comment:
This will change to wait for onmessage to be set similar to how we wait for onclose to be set.